File: src/main/java/org/trp/shincolle/client/renderer/RendererShipFishingHook.java
package org.trp.shincolle.client.renderer;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.entity.EntityRenderer;
import net.minecraft.client.renderer.entity.EntityRendererProvider;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import net.minecraft.world.entity.HumanoidArm;
import org.trp.shincolle.entity.EntityShipFishingHook;
import org.trp.shincolle.entity.base.EntityShipBase;

public class RendererShipFishingHook extends EntityRenderer<EntityShipFishingHook> {
    private static final ResourceLocation FISH_PARTICLES = ResourceLocation.withDefaultNamespace("textures/entity/fishing_hook.png");

    public RendererShipFishingHook(EntityRendererProvider.Context context) {
        super(context);
    }

    @Override
    public void render(EntityShipFishingHook entity, float yaw, float partialTicks, PoseStack poseStack, MultiBufferSource buffer, int packedLight) {
        float bobbing = Mth.cos((entity.tickCount + partialTicks) * 0.15f) * 0.05f - 0.25f;
        
        poseStack.pushPose();
        poseStack.translate(0.0D, bobbing + 0.25D, 0.0D);
        poseStack.scale(0.5f, 0.5f, 0.5f);
        poseStack.mulPose(this.entityRenderDispatcher.cameraOrientation());

        VertexConsumer vertexConsumer = buffer.getBuffer(RenderType.entityCutout(this.getTextureLocation(entity)));
        PoseStack.Pose pose = poseStack.last();
        
        vertexConsumer.addVertex(pose, -0.5f, -0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(0.0f, 1.0f).setOverlay(0).setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f);
        vertexConsumer.addVertex(pose, 0.5f, -0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(1.0f, 1.0f).setOverlay(0).setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f);
        vertexConsumer.addVertex(pose, 0.5f, 0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(1.0f, 0.0f).setOverlay(0).setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f);
        vertexConsumer.addVertex(pose, -0.5f, 0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(0.0f, 0.0f).setOverlay(0).setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f);
        poseStack.popPose();

        EntityShipBase host = entity.getHost();
        if (host != null) {
            
            double hostX = Mth.lerp(partialTicks, host.xo, host.getX());
            double hostY = Mth.lerp(partialTicks, host.yo, host.getY());
            double hostZ = Mth.lerp(partialTicks, host.zo, host.getZ());

            double hookBaseX = Mth.lerp(partialTicks, entity.xo, entity.getX());
            double hookBaseY = Mth.lerp(partialTicks, entity.yo, entity.getY());
            double hookBaseZ = Mth.lerp(partialTicks, entity.zo, entity.getZ());

            
            float bodyRot = Mth.lerp(partialTicks, host.yBodyRotO, host.yBodyRot) * (Mth.PI / 180.0F);
            double sin = Mth.sin(bodyRot);
            double cos = Mth.cos(bodyRot);

            
            double sideOffset = (host.getMainArm() == HumanoidArm.RIGHT ? 1.0D : -1.0D) * 0.35D;
            
            double forwardOffset = 0.3D;

            
            double tipX = hostX - cos * sideOffset - sin * forwardOffset;
            
            double tipY = hostY + (host.getBbHeight() * 0.45D);
            double tipZ = hostZ - sin * sideOffset + cos * forwardOffset;

            
            if (host.isCrouching()) tipY -= 0.15D;
            if (host.getIsSitting()) tipY -= host.getBbHeight() * 0.3D;

            
            double actualHookLocalY = bobbing + 0.25D;

            
            double dx = tipX - hookBaseX;
            double dy = tipY - (hookBaseY + actualHookLocalY);
            double dz = tipZ - hookBaseZ;

            VertexConsumer lineConsumer = buffer.getBuffer(RenderType.lineStrip());
            PoseStack.Pose linePose = poseStack.last();

            
            for (int i = 0; i <= 16; ++i) {
                float t = i / 16.0F;
                float lx = (float)(dx * t);
                
                float ly = (float)(actualHookLocalY + (dy * t) - (t * (1.0F - t) * 0.5F)); 
                float lz = (float)(dz * t);

                lineConsumer.addVertex(linePose, lx, ly, lz)
                        .setColor(150, 150, 150, 255) 
                        .setNormal(linePose, 0.0f, 1.0f, 0.0f);
            }
        }

        super.render(entity, yaw, partialTicks, poseStack, buffer, packedLight);
    }

    @Override
    public ResourceLocation getTextureLocation(EntityShipFishingHook entity) {
        return FISH_PARTICLES;
    }
}

File: src/main/java/org/trp/shincolle/client/renderer/RendererShipFishingHook.kt
package org.trp.shincolle.client.renderer

import com.mojang.blaze3d.vertex.PoseStack
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.entity.EntityRenderer
import net.minecraft.client.renderer.entity.EntityRendererProvider
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.Mth
import net.minecraft.world.entity.HumanoidArm
import org.trp.shincolle.entity.EntityShipFishingHook

class RendererShipFishingHook(context: EntityRendererProvider.Context) :
    EntityRenderer<EntityShipFishingHook>(context) {
    override fun render(
        entity: EntityShipFishingHook,
        yaw: Float,
        partialTicks: Float,
        poseStack: PoseStack,
        buffer: MultiBufferSource,
        packedLight: Int
    ) {
        val bobbing = Mth.cos((entity.tickCount + partialTicks) * 0.15f) * 0.05f - 0.25f

        poseStack.pushPose()
        poseStack.translate(0.0, bobbing + 0.25, 0.0)
        poseStack.scale(0.5f, 0.5f, 0.5f)
        poseStack.mulPose(this.entityRenderDispatcher.cameraOrientation())

        val vertexConsumer = buffer.getBuffer(RenderType.entityCutout(this.getTextureLocation(entity)))
        val pose = poseStack.last()

        vertexConsumer.addVertex(pose, -0.5f, -0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(0.0f, 1.0f).setOverlay(0)
            .setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f)
        vertexConsumer.addVertex(pose, 0.5f, -0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(1.0f, 1.0f).setOverlay(0)
            .setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f)
        vertexConsumer.addVertex(pose, 0.5f, 0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(1.0f, 0.0f).setOverlay(0)
            .setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f)
        vertexConsumer.addVertex(pose, -0.5f, 0.5f, 0.0f).setColor(255, 255, 255, 255).setUv(0.0f, 0.0f).setOverlay(0)
            .setLight(packedLight).setNormal(pose, 0.0f, 1.0f, 0.0f)
        poseStack.popPose()

        val host = entity.getHost()
        if (host != null) {
            val hostX = Mth.lerp(partialTicks.toDouble(), host.xo, host.x)
            val hostY = Mth.lerp(partialTicks.toDouble(), host.yo, host.y)
            val hostZ = Mth.lerp(partialTicks.toDouble(), host.zo, host.z)

            val hookBaseX = Mth.lerp(partialTicks.toDouble(), entity.xo, entity.x)
            val hookBaseY = Mth.lerp(partialTicks.toDouble(), entity.yo, entity.y)
            val hookBaseZ = Mth.lerp(partialTicks.toDouble(), entity.zo, entity.z)


            val bodyRot = Mth.lerp(partialTicks, host.yBodyRotO, host.yBodyRot) * (Mth.PI / 180.0f)
            val sin = Mth.sin(bodyRot).toDouble()
            val cos = Mth.cos(bodyRot).toDouble()


            val sideOffset = (if (host.mainArm == HumanoidArm.RIGHT) 1.0 else -1.0) * 0.35

            val forwardOffset = 0.3


            val tipX = hostX - cos * sideOffset - sin * forwardOffset

            var tipY = hostY + (host.bbHeight * 0.45)
            val tipZ = hostZ - sin * sideOffset + cos * forwardOffset


            if (host.isCrouching()) tipY -= 0.15
            if (host.isInSittingPose) tipY -= host.bbHeight * 0.3


            val actualHookLocalY = bobbing + 0.25


            val dx = tipX - hookBaseX
            val dy = tipY - (hookBaseY + actualHookLocalY)
            val dz = tipZ - hookBaseZ

            val lineConsumer = buffer.getBuffer(RenderType.lineStrip())
            val linePose = poseStack.last()


            for (i in 0..16) {
                val t = i / 16.0f
                val lx = (dx * t).toFloat()

                val ly = (actualHookLocalY + (dy * t) - (t * (1.0f - t) * 0.5f)).toFloat()
                val lz = (dz * t).toFloat()

                lineConsumer.addVertex(linePose, lx, ly, lz)
                    .setColor(150, 150, 150, 255)
                    .setNormal(linePose, 0.0f, 1.0f, 0.0f)
            }
        }

        super.render(entity, yaw, partialTicks, poseStack, buffer, packedLight)
    }

    override fun getTextureLocation(entity: EntityShipFishingHook): ResourceLocation {
        return FISH_PARTICLES
    }

    companion object {
        private val FISH_PARTICLES: ResourceLocation =
            ResourceLocation.withDefaultNamespace("textures/entity/fishing_hook.png")
    }
}
